prow/metrics/metrics.go

```diff
@@ -26,23 +26,59 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/push"
 	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/config"
 )
 
-// PushMetrics is meant to run in a goroutine and continuously push
+// NewPusher creates a Pusher which dynamically reads its config from the
+// ConfigAgent
+func NewPusher(ca configGetter) Pusher {
+	return Pusher{
+		configAgent: ca,
+	}
+}
+
+// Pusher is a struct that knows how to push metrics to prometheus
+type Pusher struct {
+	configAgent configGetter
+
+	// gatherer and waiter are only on this struct to be faked out in the tests
+	gatherer gathererFunc
+	waiter   func(time.Duration) <-chan time.Time
+}
+
+// Start is meant to run in a goroutine and continuously push
 // metrics to the provided endpoint.
-func PushMetrics(component, endpoint string, interval time.Duration) {
+func (pm Pusher) Start(component string) {
+	if pm.gatherer == nil {
```

Comment:
I'm not sure we should let test logic leak into this -- set these in `NewPusher` and put `Pusher` behind an interface if you want to ensure nobody mis-initializes it.

Reply:
This is convenient for testing. It's unlike the extra interface adds value.

Reply:
Then don't do the interface and just set the field defaults in the `NewPusher` -- IDK, seems weird to prefer the test case (where these are not nil) over the production case in the code